keps/sig-autoscaling/0001-container-resource-autoscaling.md

```diff
@@ -0,0 +1,248 @@
+
+---
+title: Container Resource based Autoscaling
+authors:
+  - "@arjunrn"
+owning-sig: sig-autoscaling
+reviewers:
+  - "@joeburnett"
+  - "@mwielgus"
+approvers:
+  - TBD
+editor: "TBD"
+creation-date: 2020-02-18
+last-updated: 2020-02-18
+status: TBD
+---
+
+# Kubernetes Enhancement Proposal Process
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade/Downgrade Strategy](#upgradedowngrade-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+The Horizontal Pod Autoscaler supports scaling of targets based on the resource usage
+of the pods in the target. The resource usage of pods is calculated as the sum
+of the individual container usage values of the pod. This is unsuitable for workloads where
+the usage of the containers are not strongly corellated or change in lockstep. This KEP
+suggests that when scaling based on resource usage the HPA also provide an option
+to consider the usages of individual containers to make scaling decisions.
+
+## Motivation
+
+An HPA is used to ensure that a scaling target is scaled up or down in such a way that the specificed
+current metric values are always maintained at a certain level. Resource based autoscaling is the most
+basic approach to autoscaling and has been present in the HPA spec since `v1`. In this mode the HPA
+controller fetches the current resource metrics for all the pods of a scaling target and then computes
+how many pods should be added or removed based on the current usage to achieve the target average usage.
+
+The target pod usage is only meaningful when the pod contains a single container. When there are
```

Comment:
> I think this is too strong of a statement.
For performance critical applications it's a requirement that usage of individual containers remain below a certain threshold so that latency guarantees are met. It's not sufficient to rely on the combined usage of all the containers in a pod to ensure performance. But I see your point and will rephrase the statement.
> I'm interested in cases where the usage does not have a direct correlation
I think I didn't do a good job with the direct correlation example. Yes the `log-shipper` will also grow but the `application` container may go from _20%_ to _60%_ where as the `log-shipper` will only go from _10%_ to _20%_ because it may be over provisioned or it's designed to sample logs and more logs don't have a direct effect. I will explain the example with more detail.